packages/gestalt-charts/src/ChartGraph/useTabularData.js

```diff
@@ -0,0 +1,139 @@
+// @flow strict-local
+import { useDefaultLabel } from 'gestalt';
+
+export type SortChangeType = 'series' | 'x' | 'y';
+export type FilterIdType = null | SortChangeType;
+export type FilterOrderType = 'desc' | 'asc';
+
+export type TransformedTabularDataType = $ReadOnlyArray<{|
+  series: string,
+  xAxis: number | string,
+  yAxis: number,
+|}>;
+
+type UseBuildCsvDataProps = ({|
+  transformedTabularData: TransformedTabularDataType,
+  isHorizontalLayout: boolean,
+|}) => string;
+
+export const useBuildCsvData: UseBuildCsvDataProps = ({
+  transformedTabularData,
+  isHorizontalLayout,
+}) => {
+  const { tableSeriesText, tableXAxisText, tableYAxisText } = useDefaultLabel('ChartGraph');
+
+  return `${tableSeriesText},${isHorizontalLayout ? tableXAxisText : tableYAxisText},${
+    isHorizontalLayout ? tableYAxisText : tableXAxisText
+  }\n${transformedTabularData
+    .map((x) => Object.values(x))
+    .map((e) => e.join(','))
+    .join('\n')}`;
+};
+
+type ElementType = {| series: string, xAxis: string, yAxis: string |};
+
+const getCompareFn = ({
+  filterId,
+  filterOrder,
+}: {|
+  filterId: FilterIdType,
+  filterOrder: FilterOrderType,
+|}) =>
+  function compareXDesc(a: ElementType, b: ElementType) {
+    let aValue = a.xAxis;
+    let bValue = b.xAxis;
+
+    if (filterId === 'y') {
+      aValue = a.yAxis;
+      bValue = b.yAxis;
+    }
+
+    if (filterId === 'series') {
+      aValue = a.series;
+      bValue = b.series;
+    }
+
+    if (filterOrder === 'desc' ? aValue < bValue : aValue > bValue) {
+      return -1;
+    }
+    if (filterOrder === 'desc' ? aValue > bValue : aValue < bValue) {
+      return 1;
+    }
+    // a must be equal to b
+    return 0;
+  };
+
+type UseTabularDataProps = ({|
+  data: $ReadOnlyArray<{|
+    name: string | number,
+    [string]: number,
+  |}>,
+  filterId: FilterIdType,
+  filterOrder: FilterOrderType,
+  tickFormatter?: {|
+    timeseries?: (number) => string | number,
+    xAxisTop?: (number, number) => string | number,
+    xAxisBottom?: (number, number) => string | number,
+    yAxisRight?: (number, number) => string | number,
+    yAxisLeft?: (number, number) => string | number,
+  |},
+  labelMap?: {| [string]: string |},
+  isHorizontalLayout: boolean,
+|}) => TransformedTabularDataType;
+
+const useTabularData: UseTabularDataProps = ({
+  data,
+  filterId,
+  filterOrder,
+  labelMap,
+  tickFormatter,
+  isHorizontalLayout,
+}) => {
+  const transformedTabularData = data
```

Comment:
nit: Same concern here. These calculations are triggered in every render wherever the hook is present. Please, correct me if I'm not understanding something.

Reply:
DONE